Counsel compile/make tweaks

stsquad opened the pull request:
Hi,

I was getting back into Linux development and realised the experience of picking up Linux targets was suboptimal especially when dealing with compile environment variables. The first patch is a simple fix that allows you to go forward. The second is a little more ambitious that allows for parsing of make help output which is fairly standard on decent sized projects. The regex could no doubt be tidied up. I don't know if it makes sense to be customizable or just there to tweak?

I'm not looking to merge it straight away but I would welcome any feedback from anyone who might care.

basil-conto commented on counsel.el:
```diff
@@ -6317,20 +6318,25 @@ list is passed to `compilation-environment'."
       ;; 0 = no-rebuild, -q & 1 needs rebuild, 2 error (for GNUMake at
       ;; least)
-      (when (< (call-process "make" nil t nil "-nqp") 2)
+      (if (> (call-process "make" nil t nil "-nqp") 1)
```
`call-process` does not always return a number.

stsquad replied:
I take it this would have failed before? What would you recommend - a cond form with multiple legs handling the various cases?

stsquad replied:
How about:

      (let ((res (call-process "make" nil t nil "-nqp")))
        (cond ((and (numberp res)
                    (< res 2))
               (goto-char (point-min))
               (while (re-search-forward counsel-compile-phony-pattern nil t)
                 (setq targets
                       (nconc targets (split-string
                                       (match-string-no-properties
                                        1))))))
              (t (setq targets '(""))))))

basil-conto replied:
> I take it this would have failed before?

Yep.

> What would you recommend - a cond form with multiple legs handling the various cases?

I would recommend `cond` whenever there is more than one condition or fall-through case.  If there is only one condition you can just use `if`.

> How about:

Passing a constant literal such as `'("")` to a destructive function such as `sort` is undefined behaviour.  Instead, you can write something like:

```el
(with-temp-buffer
  (let* ((default-directory dir)
         (res (call-process "make" nil t nil "-nqp")))
    (if (and (wholenump res) (< res 2))
        (let (targets)
          (goto-char (point-min))
          (while (re-search-forward counsel-compile-phony-pattern nil t)
            (push (split-string (match-string-no-properties 1)) targets))
          (sort (apply #'nconc targets) #'string-lessp))
      (list ""))))
```

or:

```el
(with-temp-buffer
  (let* ((default-directory dir)
         (res (call-process "make" nil t nil "-nqp"))
         targets)
    (if (or (not (numberp res)) (> res 1))
        (list "")
      (goto-char (point-min))
      (while (re-search-forward counsel-compile-phony-pattern nil t)
        (push (split-string (match-string-no-properties 1)) targets))
      (sort (apply #'nconc targets) #'string-lessp))))
```

basil-conto commented on counsel.el:
```diff
@@ -6317,20 +6318,25 @@ list is passed to `compilation-environment'."
-happen because some sort of configuration needs to be done first
-or the source tree is pristine and being used for multiple build
-trees."
+Return an single blank target (so we invoke the default target)
```
an -> a

basil-conto commented on counsel.el:
```diff
@@ -6317,20 +6318,25 @@ list is passed to `compilation-environment'."
   "Regexp for extracting phony targets from Makefiles.")
 
+(defvar counsel-compile-help-pattern
+  "\\(?:^\\(\\*?\\)[[:space:]]+\\([[:alnum:]]+\\)[[:space:]]+-\\)"
```
The first group can match the empty string, which means the `(match-beginning 1)` in `counsel-compile--probe-make-help` will always return non-`nil`.

basil-conto replied:
Change `\\(\\*?\\)` to `\\(\\*\\)?` and then `(match-beginning 1)` will return `nil` when there is no asterisk.

stsquad replied:
So I write my regex's in rx (in reb) so I had used (: bol (group (zero-or-one "*"))) for the initial part. I wonder what the rx equivalent would be to generate that?

basil-conto replied:
```el
(equal (rx bol (group (? ?*)) (+ space) (group (+ alnum)) (+ space) ?-)
       "^\\(\\*?\\)[[:space:]]+\\([[:alnum:]]+\\)[[:space:]]+-")

(equal (rx bol (? (group ?*)) (+ space) (group (+ alnum)) (+ space) ?-)
       "^\\(\\*\\)?[[:space:]]+\\([[:alnum:]]+\\)[[:space:]]+-")
```

See also https://elpa.gnu.org/packages/xr.html.

stsquad replied:
@basil-conto awesome, thanks for the heads up

basil-conto commented on counsel.el:
```diff
@@ -6379,7 +6385,44 @@ sub-directories that builds may be invoked in."
+
+(defun counsel-compile-get-make-help-invocations (&optional blddir)
+  "Have a look in the root directory for any build control files and
+query any help output.
```
Please make the docstring's first sentence fit on a single line, as per [`(info "(elisp) Documentation Tips")`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Documentation-Tips.html).

basil-conto commented on counsel.el:
```diff
@@ -6379,7 +6385,44 @@ sub-directories that builds may be invoked in."
+    (when (directory-files (or blddir srcdir) nil
+                           counsel-compile-make-pattern t)
+      (counsel--compile-get-make-targets
+       'counsel-compile--probe-make-help srcdir blddir))))
```
Nit: Quote function symbols with `#'` instead of `'`.

basil-conto commented on counsel.el:
```diff
@@ -6379,7 +6385,44 @@ sub-directories that builds may be invoked in."
                            counsel-compile-make-pattern t)
-      (counsel--compile-get-make-targets srcdir blddir))))
+      (counsel--compile-get-make-targets
+       'counsel-compile--probe-make-targets srcdir blddir))))
```
Nit: Quote function symbols with `#'` instead of `'`.

basil-conto commented on counsel.el:
```diff
@@ -6379,7 +6385,44 @@ sub-directories that builds may be invoked in."
+    ;; only proceed if the help target exists
+    (when (= (call-process "make" nil t nil "-q" "help") 1)
+      (with-temp-buffer
+        (when (= (call-process "make" nil t nil "help") 0)
```
`call-process` does not always return a number.

basil-conto commented on counsel.el:
```diff
@@ -6379,7 +6385,44 @@ sub-directories that builds may be invoked in."
+        (primary-targets nil)
+        (targets nil))
+    ;; only proceed if the help target exists
+    (when (= (call-process "make" nil t nil "-q" "help") 1)
```
Should the `DESTINATION` argument of `call-process` really be `t` here? If so, please add a comment explaining why this is the case, as it is not clear where the output (if any) will be inserted.

basil-conto commented on counsel.el:
```diff
@@ -6379,7 +6385,44 @@ sub-directories that builds may be invoked in."
+          (goto-char (point-min))
+          (while (re-search-forward counsel-compile-help-pattern nil t)
+            (let ((text (match-string-no-properties 2)))
+            (if (match-beginning 1)
```
Indentation seems off here. See also my previous comment about the value of `(match-beginning 1)` here.

basil-conto commented on counsel.el:
```diff
@@ -6379,7 +6385,44 @@ sub-directories that builds may be invoked in."
+                (setq primary-targets (nconc primary-targets (list text)))
+              (setq targets (nconc targets (list text))))))))
+      (append (sort primary-targets #'string-lessp)
+              (sort targets #'string-lessp)))))
```
It is almost always more efficient to do something like:

```el
(let (elems)
  (while ...
    (push elem elems))
  (nreverse elems))
```

instead of appending each element to the list each time.

Since you are destructively sorting the lists at the end, however, the order of elements does not matter, so you can just write something like:

```el
(let (primary-targets targets)
  ...
  (push text (if (match-beginning 1) primary-targets targets))
  ...
  (nconc (sort primary-targets #'string-lessp)
         (sort targets #'string-lessp)))
```

stsquad replied:
Hmm I've fixed it up as suggested but I also needed to add env support (for example the linux help output is affected by the compile environment). I wonder if this needs a bit of refactoring?

    (defun counsel-compile--probe-make-help (dir)
      "Return a list of Make targets based on help for `DIR'.

    It is quite common for a 'make help' invocation to return a human
    readable list of targets.  Often common targets are marked with a
    leading asterisk.  The exact search pattern is controlled by
    `counsel-compile-help-pattern'."
      (let* ((default-directory dir)
             (c-env (split-string
                     (substring-no-properties
                      (mapconcat #'identity counsel-compile-env " "))))
             (p-res (apply #'call-process "make" nil nil nil "-q" "help" c-env)))
        ;; only proceed if the help target exists
        (when (and (numberp p-res) (= p-res 1))
          (with-temp-buffer
            (let (primary-targets targets)
              (setq p-res
                    (apply #'call-process "make" nil t nil "help" c-env))
              (when (and (numberp p-res) (= p-res 0))
                (goto-char (point-min))
                (while (re-search-forward counsel-compile-help-pattern nil t)
                  (let ((text (match-string-no-properties 2)))
                    (push text
                          (if (string= "*" (match-string-no-properties 1))
                              primary-targets targets)))))
              (nconc (sort primary-targets #'string-lessp)
                     (sort targets #'string-lessp)))))))

basil-conto replied:
> I wonder if this needs a bit of refactoring?

Let me know if my comments address your concern.

stsquad replied:
TIL eql, handy. Looking good

basil-conto replied:
> TIL eql

For fixnums you could equivalently use `eq`, but in general this is considered passé and `eql` is preferred instead, especially since the latter works with bignums in Emacs 27 and above:

```el
emacs-version
;; => "27.0.91"

(eq (1+ most-positive-fixnum)
    (1+ most-positive-fixnum))
;; => nil

(eql (1+ most-positive-fixnum)
     (1+ most-positive-fixnum))
;; => t
```

stsquad commented:
Thanks for you review @basil-conto - I've pushed a new revision which addresses all your comments although I'm wondering if counsel-compile--probe-make-help could be improved to be a bit less ugly.

There is also another commit which fixes up M-i to be more aware of counsel-compile and give a cleaner experience than the default ivy editing.

basil-conto commented on counsel.el:
```diff
@@ -6356,12 +6361,13 @@ text with FACE."
-(defun counsel--compile-get-make-targets (srcdir &optional blddir)
-  "Return a list of Make targets for a given SRCDIR/BLDDIR combination.
+(defun counsel--compile-get-make-targets (probe-fn srcdir &optional blddir)
+  "Return propertied make targets returned by `PROBE-FN' in `SRCDIR'.
```
propertied -> propertized

Also, metavariables here and elsewhere should not be surrounded by backtick/apostrophe; see [`(info "(elisp) Documentation Tips")`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Documentation-Tips.html).

basil-conto commented on counsel.el:
```diff
@@ -6388,7 +6394,47 @@ sub-directories that builds may be invoked in."
+      (counsel--compile-get-make-targets
+       #'counsel-compile--probe-make-targets srcdir blddir))))
+
+
```
Nit: extraneous newline.

basil-conto commented on counsel.el:
```diff
@@ -6388,7 +6394,47 @@ sub-directories that builds may be invoked in."
+                    (if (string= "*" (match-string-no-properties 1))
+                        primary-targets targets)))
+            (nconc (sort primary-targets #'string-lessp)
+                   (sort targets #'string-lessp))))))))
```
Combined with my previous comment about `match-beginning`, can this be simplified as follows?

```el
(let ((default-directory dir)
      primary-targets targets)
  ;; Only proceed if the help target exists.
  (when (eql 1 (apply #'call-process "make" nil nil nil "-q" "help"
                      counsel-compile-env))
    (with-temp-buffer
      (when (eql 0 (apply #'call-process "make" nil t nil "help"
                          counsel-compile-env))
        (goto-char (point-min))
        (while (re-search-forward counsel-compile-help-pattern nil t)
          (push (match-string 2)
                (if (match-beginning 1) primary-targets targets)))
        (nconc (sort primary-targets #'string-lessp)
               (sort targets #'string-lessp))))))
```

stsquad replied:
@basil-conto: do properties just disappear when strings are sent outside of Emacs?

basil-conto replied:
> do properties just disappear when strings are sent outside of Emacs?

Text properties are a mostly transparent Emacs-specific data representation, so yes.

Even within Emacs, two strings with the same contents but different properties are still `equal`.

In general you shouldn't need to worry about application-specific properties like the ones Counsel uses, as they have no visible effects on other parts of Emacs (unless there's a clash of property names).

stsquad replied:
That code tested beautifully ;-)

basil-conto commented on counsel.el:
```diff
@@ -6502,16 +6548,46 @@ specified by the `blddir' property."
+  "Insert current compile command into mini-buffer for editing.
+
+This mirrors the behaviour of `ivy-insert-current' but with specific
+handling for the counsel-compile metadata."
```
Nits:
* "Minibuffer" isn't usually hyphenated.
* Ivy follows the Emacs convention of US spelling, so behaviour -> behavior.
* Quote `counsel-compile` with backtick/apostrophe.

basil-conto commented on counsel.el:
```diff
@@ -6502,16 +6548,46 @@ specified by the `blddir' property."
+handling for the counsel-compile metadata."
+  (interactive)
+  (delete-minibuffer-contents)
+  (let* ((end (and ivy--directory
```
`end` doesn't seem to be used?

basil-conto replied:
`end` still doesn't seem to be used.

basil-conto commented on counsel.el:
```diff
@@ -6502,16 +6548,46 @@ specified by the `blddir' property."
+         (blddir (get-text-property 0 'blddir cmd)))
+    (when blddir
+      (setq counsel-compile--current-build-dir blddir))
+    (if (get-char-property 0 'cmd cmd)
```
`get-char-property` -> `get-text-property`?

basil-conto commented on counsel.el:
```diff
@@ -6502,16 +6548,46 @@ specified by the `blddir' property."
+    (if (get-char-property 0 'cmd cmd)
+        (insert (substring-no-properties
+                 cmd 0 (next-single-property-change 0 'cmd cmd)))
+      (insert (substring-no-properties cmd)))))
```
AKA?

```el
(insert (substring-no-properties
         cmd 0 (and (get-text-property 0 'cmd cmd)
                    (next-single-property-change 0 'cmd cmd))))
```

basil-conto commented on counsel.el:
```diff
@@ -6502,16 +6548,46 @@ specified by the `blddir' property."
+;; operation which doesn't include the metadata we want.
+(defvar counsel-compile-map
+  (let ((map (make-sparse-keymap)))
+    (define-key map (kbd "M-i") #'counsel-compile-edit-command)
```
Would `[remap ivy-insert-current]` work? And would it be any better? (I don't know.)

basil-conto commented on counsel.el:
```diff
@@ -6502,16 +6548,46 @@ specified by the `blddir' property."
+  (let ((map (make-sparse-keymap)))
+    (define-key map (kbd "M-i") #'counsel-compile-edit-command)
+    map)
+  "Additional ivy keybindings during command selection")
```
Missing full stop.

basil-conto commented on counsel.el:
```diff
@@ -6502,16 +6548,46 @@ specified by the `blddir' property."
-  "Call `compile' completing with smart suggestions, optionally for DIR."
+  "Call `compile' completing with smart suggestions, optionally for DIR.
+
+Additional actions:\\<ivy-minibuffer-map>"
```
On its own, `\\<...>` is a no-op. Did you mean the following instead?

```
Additional actions:

\\{ivy-minibuffer-map}
```

stsquad replied:
Actually I think I'll drop this - it just adds noise to the help.

stsquad commented:
OK I have just pushed v3 to the branch.

basil-conto commented:
Just in case this gets hidden in GitHub's UI - my only remaining comment is that the local binding for `end` in `counsel-compile-edit-command` does not seem to be used.

stsquad commented:
> Just in case this gets hidden in GitHub's UI - my only remaining comment is that the local binding for `end` in `counsel-compile-edit-command` does not seem to be used.

Oops sorry I missed that. Left over code from the original ivy-insert-current. Fixed in v4 I've just pushed.

basil-conto commented:
Thanks, LGTM.

stsquad commented:
Gentle ping to @abo-abo - is there anything else you would like to see for this?

stsquad commented:
Removed the #2652 fix from the PR branch so it doesn't get accidentally included.

abo-abo commented:
Merged, thanks.

One thing that can maybe be improved is here:

```elisp
(setq counsel-compile--current-build-dir blddir)
```

Maybe it's possible to have the build dir be more explicit, e.g. `cd src && make -k all`?
Then instead of `counsel-compile-edit-command`, we could still use `ivy-insert-current`.
